Add initial (cleaned up) implementations

gfontenot opened the pull request:
These are _almost_ pulled directly from my dotfiles. I did do a little cleanup
to reorganize them under the `xcode` wrapper, and I gave all of them help
commands, but they work well as they are and are battle tested on my machine
fairly well.

gfontenot committed:
Add initial (cleaned up) implementations

These are _almost_ pulled directly from my dotfiles. I did do a little cleanup
to reorganize them under the `xcode` wrapper, and I gave all of them help
commands, but they work well as they are and are battle tested on my machine
fairly well.

sharplet commented on src/commands/xcode-list:
```diff
@@ -0,0 +1,53 @@
+    ;;
+esac
+
+apps=$(find /Applications -maxdepth 1 -name "Xcode*.app")
```
This doesn't handle paths with spaces: thoughts on using a Bash array here? e.g., `apps=(/Applications/Xcode*.app)`

gfontenot replied:
I'm trying to stay away from bash for reasons I can't really justify. This file is actually already failing some shellcheck stuff for POSIX shell so I'm going to rework it a bit.

sharplet reviewed (approved):
Nice 👍 

I noticed a potential issue around handling spaces. Also wondering: should the subcommands go in a `libexec` folder? I don't have strong opinions around that, but it's a thing I've seen other tools do.

gfontenot committed:
xcode-list: Fix support for spaces in filenames

Previously, if we had `/Applications/Xcode 4.app`, we'd end up with broken
output:

```
❯ xcode-list
Xcode
4
Xcode-8.3.3
Xcode-9-beta-5
```

With this change, we've now made this file POSIX compliant, while also
providing better support for files with spaces in them:

```
❯ xcode-list
Xcode 4
Xcode-8.3.3
Xcode-9-beta-5
```

gfontenot commented:
I don't know where the "correct" place for the subcommands to live, tbh. In my tests, I've been installing everything to `/usr/local/bin` although maybe that's not exactly correct either? This seems like (it might be) a source organization thing rather than a semantic user-facing thing so I think I'm OK leaving them where they currently are?

sharplet commented:
Yeah it's not a big deal, I was mostly just wondering out loud if there's a conventional place to put that stuff.

sharplet commented on src/commands/xcode-list:
```diff
@@ -0,0 +1,49 @@
+esac
+
+if [ "$full_path" -eq "1" ]; then
+  find /Applications -maxdepth 1 -name "Xcode*.app"
```
You could potentially extract a `find_xcodes` function for these common arguments if you wanted to DRY this up slightly.

gfontenot committed:
xcode-list: DRY up the command to find Xcode apps

Instead of repeating this, we can extract a small function out that knows how
to perform the find operation and also passes along any extra flags. This lets
us use a single definition of the find command.

Unfortunately, this breaks shellcheck's ability to parse the flags passed to
find, and so we get some warnings about $1 not being expanded. That's fine,
and we know that, so we'll add a small comment and tell shellcheck to ignore
that warning for the line.

gabebw commented on README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Extending
+
+This tool follows the same pattern set by `git` and similar commands. It
```
Innnnteresting! I like it!

gabebw commented on src/commands/xcode-list:
```diff
@@ -0,0 +1,49 @@
+error(){
+  red="\e[1;31m"
+  reset="\e[0m"
+  printf "${red}%s${reset}\n" "$1"  >&2
```
Nitpick: extra space before `>`

gabebw commented on src/commands/xcode-open:
```diff
@@ -0,0 +1,53 @@
+
+app=$(xcode-version --path)
+
+case $1 in
```
Why not use getopts? Just didn't want to deal with it. (I wouldn't blame you.)

gfontenot replied:
Exactly. I also want to provide long switches for the sake of readability/scripting.

gabebw commented on src/commands/xcode-switch:
```diff
@@ -0,0 +1,49 @@
+
+if other_apps=$(xcode-list --full-path | grep --invert-match "$current_app"); then
+  echo "$other_apps" \
+    | fzf --query="$1" --select-1 --print0 \
```
Worth checking that fzf is installed at the top of this script? `command -v fzf >/dev/null` is cross-platform, i think.

gfontenot replied:
We're going to declare `fzf` as a dependency in the homebrew formula so that we can rely on it. I think long-term I'd like to make this configurable, and also provide a fallback for if no fzf-like tool is available.

gabebw commented on src/xcode:
```diff
@@ -0,0 +1,54 @@
+  echo "$PATH" \
+  | tr ":" "\n" \
+  | xargs -I {} sh -c 'test -d {} && echo {}' \
+  | xargs -I {} find {} -maxdepth 1 -perm +111 -iname "xcode-*" \
```
Why `iname` instead of `name`? The code on line 43 checks against `xcode` (lowercase), so I think having a command named `XCoDe-foo` in the path wouldn't work in any case.

gfontenot replied:
Good point

gfontenot committed:
Remove extra spaces before redirect

I _really_ need to figure out a good way to share this code across scripts,
instead of dumping this into the top of each file.

gfontenot committed:
xcode: Use case sensitive comparison for find

We're doing a case sensitive comparison below this, so it's probably best to
just use a case sensitive comparison everywhere. This will provide a more
consistent experience.
